src/NHibernate.Test/Criteria/EntityProjectionsTest.cs

```diff
@@ -0,0 +1,448 @@
+				{
+					Key = new CompositeKey
+					{
+						Id1 = 1,
+						Id2 = 2
+					},
+					Name = "Composite"
+				};
+
+				session.Save(child1);
+				session.Save(child2);
+				session.Save(parent.SameTypeChild);
+				session.Save(parent);
+				session.Save(_entityWithCompositeId);
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public void RootEntityProjectionFullyInitializedAndWithUnfetchedLazyPropertiesByDefault()
+		{
+			using (var sqlLog = new SqlLogSpy())
+			using (var session = OpenSession())
+			{
+				Sfi.Statistics.Clear();
+				EntityComplex entityRoot = session
+					.QueryOver<EntityComplex>()
+					.Where(ec => ec.LazyProp != null)
+					.Select(Projections.RootEntity())
+					.Take(1).SingleOrDefault();
+
+				Assert.That(NHibernateUtil.IsInitialized(entityRoot), Is.True, "Object must be initialized by default");
+				Assert.That(session.IsReadOnly(entityRoot), Is.False, "Object must not be readonly by default");
+				Assert.That(NHibernateUtil.IsPropertyInitialized(entityRoot, nameof(entityRoot.LazyProp)), Is.False, "Lazy properties should not be initialized by default.");
+				Assert.That(sqlLog.Appender.GetEvents().Length, Is.EqualTo(1), "Only one SQL select is expected");
+			}
+		}
+
+		[Test]
+		public void RootEntityProjectionLazy()
+		{
+			using (var session = OpenSession())
+			{
+				EntityComplex entityRoot = session
+					.QueryOver<EntityComplex>()
+					.Select(Projections.RootEntity().SetLazy(true))
+					.Take(1).SingleOrDefault();
+
+
+				Assert.That(NHibernateUtil.IsInitialized(entityRoot), Is.False, "Object must be lazy loaded.");
+			}
+		}
+
+		[Test]
+		public void AliasedEntityProjection()
+		{
+			using (var sqlLog = new SqlLogSpy())
+			using (var session = OpenSession())
+			{
+				EntitySimpleChild child1 = null;
+				child1 = session
+					.QueryOver<EntityComplex>()
+					.JoinAlias(ep => ep.Child1, () => child1)
+					.Select(Projections.Entity(() => child1))
+					.Take(1).SingleOrDefault<EntitySimpleChild>();
+
+				Assert.That(child1, Is.Not.Null);
+				Assert.That(NHibernateUtil.IsInitialized(child1), Is.True, "Object must be initialized");
+				Assert.That(sqlLog.Appender.GetEvents().Length, Is.EqualTo(1), "Only one SQL select is expected");
+			}
+		}
+
+		[Test]
+		public void EntityProjectionAsSelectExpression()
+		{
+			using (var sqlLog = new SqlLogSpy())
+			using (var session = OpenSession())
+			{
+				EntitySimpleChild child1 = null;
+				child1 = session
+					.QueryOver<EntityComplex>()
+					.JoinAlias(ep => ep.Child1, () => child1)
+					.Select(ec => child1.AsEntity())
+					.Take(1).SingleOrDefault<EntitySimpleChild>();
+
+				Assert.That(child1, Is.Not.Null);
+				Assert.That(NHibernateUtil.IsInitialized(child1), Is.True, "Object must be initialized");
+				Assert.That(sqlLog.Appender.GetEvents().Length, Is.EqualTo(1), "Only one SQL select is expected");
+			}
+		}
+
+		[Test]
+		public void EntityProjectionLockMode()
```

Comment:
This one fails for Oracle. It generates following query:
```sql
select id1_1_0_,name2_1_0_
    from (
        SELECT child1x1_.Id as id1_1_0_, child1x1_.Name as name2_1_0_
            FROM EntityComplex this_
            inner join EntitySimpleChild child1x1_
                on this_.Child1Id=child1x1_.Id )
    where rownum <=:p0
    for update of child1x1_.Id
```
It fails with error:
```
ORA-00904: "CHILD1X1_"."ID" : invalid identifier
```

The for update clause is bad. I think this is an unrelated bug, indeed #1352.
So better add something like
```c#
if (Dialect is Oracle8iDialect)
	Assert.Ignore("Oracle is not supported due to #1352 bug (NH-3902)");
```